### ISSUE TITLE
Initial Design 

### ISSUE THREAD

**Issue by alexnederlof** [Magnetme/task-scheduler#1]
We want to schedule jobs periodically and incidentally, on a distributed system, and be fault-tolerant. Googling led to know re-usable existing solutions, so lets build our own, but only build the things we absolutely need, and make it awesome. 

# Prerequisites 

- I'll call the service/node executing a job the *executor*. I'll call this service *the scheduler*. The scheduler consists of redundant nodes.
- I assume Kafka and Zookeeper always available.
- When PostgreSQL goes down this can be seen as "all schedulers go down".

# Problem:

1. We want to run some jobs periodically 
2. It should not matter which application, node, server, internet-of-things-refrigerator, or whatever executes this job (e.g. location unaware) 
3. It should be fault-tolerant in the following scenarios:
    1. The executing node aborts and lets the scheduled know
    2. The executing node disappears and the scheduler needs to find out
    3. The scheduler is taken down
4. It should run on the Magnet.me stack (Too bad, no [Chronos](https://www.youtube.com/watch?v=FLqURrtS8IA) or [Oozie](http://oozie.apache.org/),  for us)
5. It should allow incidental kickoff of jobs (e.g. let a developer run something form a UI) and scheduled repetitive jobs.

# Design:

A Java/Jersey/Angular Microservice that uses PostgreSQL as a storage, and Zookeeper for locking. 

The postgres table might look like this:

name  | type | goal
-------- | ------ | ------
id | number | Random ID of this scheduled job
job_id | string | Id of the job. For example "calculate_churn_metrics"
created_at | date | When the job was scheduled (either by a human or because it is recurring)
execute_at | date | When it should be executed
request_started_at | date | The date a request for execution was sent.
started_at | date | When it received ACK that the task is being executed
started_by | string | The id sent by the node executing the task, to identify itself. 
completed_at | optional Date | when did it complete
progress | optional int [0...100] | if the job pings back it's progress, it is saved here
last_ping | optional date | The last time we heard from the job. Either by ping or by progress report. This is used to determine if a job is lost.
completion_status | enum { completed, failed, aborted, lost } | The state at completion, if the job was completed
completion_message | text | Arbitrary text field to ~~hack in everything we forgot in the initial design~~ report any message like "Updated 10 items". Only meant logging purposes.

And another table for recurring jobs, that configures how often a job should run

name  | type | goal
-------- | ------ | ------
job_id | string | Id of the job. For example "calculate_churn_metrics"
interval | enum { hourly, daily, weekly } | How often it should run
parallel | boolean | should it await completion of its predecessor? 
interval | boolean | if true, run jobs at a fixed interval (every hour). If false, run them [interval amount of time] after the last completion.
reschedule_on_abort | boolean | 
reschedule_on_failure | boolean | 
reschedule_cool_down | int | Amount of time to wait before rescheduling

# How it works

### Scheduling jobs:

1. The service checks Zookeeper (with Curator) to see if it is the master. 
2. If it is, it polls for any jobs that should be executed now (e.g. `run_at < NOW()`) 
3. For every job that needs to run
    1. Send event over Kafka that requests its execution
    2. Set `request_started_at = NOW()`
4. Goto 1.

### Polling lost jobs 

This is not in the MVP, because it won't happen that often but:

1. Check for jobs that have not pinged back in the last X amount of time
2. Rescheduled them if necessary.

### The rest of the work is listening to Kafka:

1. Upon **job_started**
    1. Set `started_at = NOW(); started_by = id` for the Job
1. Upon**job_progress** or **job_ping**: Set `progress = progress` for the Job
1. Upon **job_finished**: Set the completion_status of the finished job
    1. If status `success` then nothing
    1. If other status, then check if the job requires a re-run. If so, re-scheduled the job after cooldown. Otherwise no_op.

# Disaster strikes.

### When the executor goes down.

The server receives no pings and progress anymore. After a determined or derived amount of time we can re-schedule the service (this is not in the MVP).

### When a scheduler node goes down

Another scheduler node becomes master and starts dispatching jobs. If all scheduler nodes go down it will simply "catch up" when they come back online. 

Later on we might implement a flag that disallows this behavior for certain jobs, in case it is "time of the day" sensitive.

**Comment by michaeldejong** [comment]
For what kind of jobs do you expect this scheduler to be used?

I can imagine that for most services you'd want them to manage their own background tasks themselves. Thinking of OAuth, there are several clean-up jobs running in the background at any time, but I don't think those should be "outsourced". Perhaps you're thinking of different kinds of jobs?

**Comment by AlexKolpa** [comment]
Well, Oauth is a pretty big service already, but I can imagine we don't want to 'pollute' the smalller services with a scheduler. This is already something I ran into with my dispatcher service, which needs to notify users at certain times. Getting these job notifications over Kafka would be much simpler than setting up a queuing system in the service itself.

**Comment by michaeldejong** [comment]
Well I don't think OAuth is a particularly big service. It's still maintained by one person which fits the microservice mantra. 

I believe we never decided on which level we want to create microservices. We could go for either a business component size ie.: user management, authentication, messaging, networks, pep, etc. Or you could go for the technical component size ie.: scheduler, mailer, etc. Perhaps this is not the place to have this discussion, but I do feel every service should be as standalone as possible which is more in-line with the business component size.

**Comment by AlexKolpa** [comment]
Good point about being overly reliant on other services. That's something that might become an issue in the future.

**Comment by rogierslag** [comment]
I want also an `active` flag, to temporarily disable services while performing maintenance.

Since we process the stuff over kafka the direct coupling between the services is not that high. I agree that doing some tasks with this is easier than let everyone spawn executors for just in case behaviour.

Finally I'm thinking about placing this data in ES. It's completely redundant, and easily manageable in this case. It also kills the SPOF we have going on

**Comment by alexnederlof** [comment]
The reason I put this in a separate service, is that I think we will want to re-use the logic to do this kind of task scheduling. For example, in Magnet.me it will schedule

1. Certain re-index jobs
2. Periodic metrics
3. Clean-up jobs
4. Periodic emails.

Now a service (like OAuth) can choose to do it itself, which is fine. But if you want more features like guarantees, distribution, etc, you can use this microservice. Now you could build this into every service we write, or we can do it like this. Since the problem is pretty standard (and abstract) I thought a separate service would be of added value here.

**Comment by alexnederlof** [comment]
To be clear; this service just sends events. It does not actually perform the task itself.

**Comment by TiddoLangerak** [comment]
Couple of thoughts:
- I do like having this as a stand-alone microservice since it would make it a lot easier to deal with scheduled tasks in distributed microservices as well as makes it easier to schedule tasks across microservices.
- You didn't mention it explicitly, but I assume that microservices can register jobs themselves over an api?
- I think it would be a good idea to add an `owner` field to the jobs, such that we can easily query for all jobs for a certain microservice.
- It seems pretty inefficient to continuously poll the database for tasks. Perhaps we can better do this reactively, i.e. schedule jobs (in memory) on startup/leader election and update this list live based on API requests.

**Comment by alexnederlof** [comment]
> It seems pretty inefficient to continuously poll the database for tasks. Perhaps we can better do this reactively, i.e. schedule jobs (in memory) on startup/leader election and update this list live based on API requests.

Totally agree! But not in the MVP :wink:

**Comment by erwinvaneyk** [comment]
I really like the concept behind this. However I don't like it being dependent on postgres. I know we generally don't want additional database systems, but this service only requires minor storage. To me it seems decoupling it from the main database is bigger advantage than keeping one central database.

**Comment by michaeldejong** [comment]
> To me it seems decoupling it from the main database is bigger advantage than keeping one central database. 

Well nobody said it would run in the same database as magnet. It might run on the same database server, but I see no problem with that.

**Comment by erwinvaneyk** [comment]
Ah okay, I read it as implied that it would run on the same database(server).

**Comment by rogierslag** [comment]
> It should not matter which application, node, server, internet-of-things-refrigerator, or whatever executes this job (e.g. location unaware)

How will we incorporate the differences between eg Java services and NodeJS services? These cannot simply run anywhere. We can ping the client that is should execute it, but in case of a failure this means that that client cannot run the job (so we need a list of clients, which need to be populated from eg puppet). This kinda further complicates the concept

**Comment by michaeldejong** [comment]
> We can ping the client that is should execute it, but in case of a failure this means that that client cannot run the job (so we need a list of clients, which need to be populated from eg puppet).

You mean like the client has become unavailable? Does that really matter? The event gets pushed onto Kafka, and the client will see it when it comes back online. For redundancy you could run multiple instances of the same version...

**Comment by alexnederlof** [comment]
Exactly, if the scheduler gets no ACK, then the service (in whatever language) was not available, so it runs the appropriate recovery.

**Comment by rogierslag** [comment]
derp for some reason I suddenly had in my head the scheduler would also execute the code instead of pinging the service to execute it. herpderp sorry

Another more valid point: can we also add arguments to the job table? That way we can easily let the service in question do a specific thing, instead of having separate endpoints and parsing for every single thing

**Comment by alexnederlof** [comment]
Sure, but not MVP

**Comment by rogierslag** [comment]
:+1:
